spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/ReactiveRedisStreamMessageProducer.java

```diff
@@ -0,0 +1,228 @@
+		this.consumerName = consumerName;
+	}
+
+	/**
+	 * Create the Consumer Group if and only if it does not exist. During the
+	 * creation we also create the stream, see {@code MKSTREAM}.
+	 * @param createConsumerGroup specify if we should create the Consumer Group, {@code false} by default
+	 */
+	public void setCreateConsumerGroup(boolean createConsumerGroup) {
+		this.createConsumerGroup = createConsumerGroup;
+	}
+
+	/**
+	 * Set {@link StreamOperations} used to customize the {@link StreamReceiver}.
+	 * It provides a way to set the polling timeout and the serialization context.
+	 * By default the polling timeout is set to infinite and {@link StringRedisSerializer} is used.
+	 * @param streamReceiverOptions the desired receiver options
+	 * */
+	public void setStreamReceiverOptions(@Nullable StreamReceiver.StreamReceiverOptions<String, ?> streamReceiverOptions) {
+		this.streamReceiverOptions = streamReceiverOptions;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "redis:stream-inbound-channel-adapter";
+	}
+
+	@Override
+	protected void onInit() {
+		super.onInit();
+		this.streamReceiver = StreamReceiver.create(this.reactiveConnectionFactory, this.streamReceiverOptions);
+		if (StringUtils.hasText(this.consumerName) && StringUtils.isEmpty(this.consumerGroup)) {
+			this.consumerGroup = getBeanName();
+		}
+		this.reactiveRedisTemplate = new ReactiveRedisTemplate<>(this.reactiveConnectionFactory,
+				RedisSerializationContext.string());
+		this.reactiveStreamOperations = this.reactiveRedisTemplate.opsForStream();
+	}
+
+	@Override
+	protected void doStart() {
+		super.doStart();
+
+		StreamOffset<String> offset = StreamOffset.create(this.streamKey, this.readOffset);
+
+		Flux<Message<?>> messageFlux;
+
+		if (StringUtils.isEmpty(this.consumerName)) {
+			messageFlux = this.streamReceiver
+					.receive(offset)
+					.map(event -> getMessageBuilderFactory().withPayload(this.extractPayload ? event.getValue() : event)
+							.setHeader(RedisHeaders.STREAM_KEY, event.getStream())
+							.setHeader(RedisHeaders.STREAM_MESSAGE_ID, event.getId())
+							.build());
+		}
+		else {
+			Assert.notNull(this.consumerGroup, "'consumerGroup' must not be null");
+
+			if (this.createConsumerGroup) {
+				Assert.hasText(this.consumerName, "'consumerName' must be set");
+				//We check that the group does not already exist before creating it
+				this.reactiveRedisTemplate.hasKey(this.streamKey)
+						.subscribe(keyExists -> {
+							if (!keyExists) {
+								this.reactiveStreamOperations.createGroup(this.streamKey, this.consumerGroup)
+										.subscribe();
+							}
+							else {
+								this.reactiveStreamOperations.groups(this.streamKey)
+										.map(StreamInfo.XInfoGroup::groupName)
+										.filter(groupName -> groupName.equals(this.consumerGroup))
+										.switchIfEmpty(this.reactiveStreamOperations.createGroup(this.streamKey,
+												this.consumerGroup))
+										.subscribe();
+							}
```

Comment:
Why so complex logic?
Can't we just call `this.reactiveStreamOperations.createGroup()`?
And also I asked to not call `subscribe()`.
There is `Mono.thenMany()` for our reactive stream.

Reply:
`this.reactiveStreamOperations.createGroup()` throws an exception if the group already exists:
```
Caused by: io.lettuce.core.RedisBusyException: BUSYGROUP Consumer Group name already exists
	at io.lettuce.core.internal.ExceptionFactory.createExecutionException(ExceptionFactory.java:126) ~[lettuce-core-6.0.0.M1.jar:6.0.0.M1]
```
That's why I try to verify it following this logic: 
 ```
  IF streamKey NOT exist
       CREATE the group and the stream
  ELSE 
       CHECK the groups subscribed to the stream // XinfoGroup
            IF the group is not absent 
                CREATE it
```

Reply:
If group already exists, and an exception is thrown, we always can fallback to the rest of the stream using `onErrorResume()`.
I understand that this is not what language does for us, but I feel like with just Reactor operators the logic should be simpler and we just do less calls over the network when group really doesn't exist.

Reply:
Sure, but I lack experience on Reactor.
Learning Reactor is in my TODO list.

Reply:
No problem! 
I can clean that up for you during merge.
Let's nail all the other concerns and we are good to go!